eth/plasma.go

```diff
@@ -0,0 +1,295 @@
+	plasmaContract, err := contracts.NewPlasmaMVP(contractAddr, client.ec)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a session with the contract and operator account
+	auth := bind.NewKeyedTransactor(privateKey)
+	plasmaSession := &contracts.PlasmaMVPSession{
+		Contract: plasmaContract,
+		CallOpts: bind.CallOpts{
+			Pending: true,
+		},
+		TransactOpts: bind.TransactOpts{
+			From:     auth.From,
+			Signer:   auth.Signer,
+			GasLimit: 3141592, // aribitrary. TODO: check this
+		},
+	}
+
+	// TODO: deal with syncing issues
+	lastCommittedBlock, err := plasmaSession.LastCommittedBlock()
+	if err != nil {
+		return nil, fmt.Errorf("Contract session not correctly established - %s", err)
+	}
+
+	plasma := &Plasma{
+		session: plasmaSession,
+		client:  client,
+		// capacity argument is advisory and not enforced in the memdb implementation
+		// TODO: flush the in-memory DB to a local one to bound memory consumption
+		memdb:  memdb.New(comparer.DefaultComparer, 1),
+		logger: logger,
+
+		ethBlockNum:   big.NewInt(-1),
+		blockNum:      lastCommittedBlock,
+		finalityBound: finalityBound,
+
+		lock: &sync.Mutex{},
+	}
+
+	// listen to new ethereum block headers
+	ethCh, err := client.SubscribeToHeads()
+	if err != nil {
+		logger.Error("Could not successfully subscribe to heads: %v", err)
+		return nil, err
+	}
+
+	// start listeners
+	go watchEthBlocks(plasma, ethCh)
+	go watchDeposits(plasma)
+	go watchExits(plasma)
+
+	return plasma, nil
+}
+
+// SubmitBlock proxy. TODO: handle batching with a timmer interrupt
+func (plasma *Plasma) SubmitBlock(header [32]byte, numTxns *big.Int, fee *big.Int) (*types.Transaction, error) {
+	plasma.blockNum = plasma.blockNum.Add(plasma.blockNum, big.NewInt(1))
+
+	tx, err := plasma.session.SubmitBlock(
+		[][32]byte{header},
+		[]*big.Int{numTxns},
+		[]*big.Int{fee},
+		plasma.blockNum)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return tx, nil
+}
+
+// GetDeposit checks the existence of a deposit nonce
+func (plasma *Plasma) GetDeposit(nonce *big.Int) (*plasmaTypes.Deposit, error) {
+	key := prefixKey(depositPrefix, nonce.Bytes())
+	data, err := plasma.memdb.Get(key)
+
+	var deposit plasmaTypes.Deposit
+	// check against the contract if the deposit is not in the cache or decoding fails
+	if err != nil || json.Unmarshal(data, &deposit) != nil {
+		if plasma.memdb.Contains(key) {
+			plasma.logger.Info("corrupted deposit found within db")
+			plasma.memdb.Delete(key)
+		}
+
+		d, err := plasma.session.Deposits(nonce)
+		if err != nil {
+			plasma.logger.Error("contract call, deposits, failed")
+			return nil, err
+		}
+
+		if d.CreatedAt.Sign() == 0 {
+			return nil, fmt.Errorf("deposit does not exist")
+		}
+
+		deposit = plasmaTypes.Deposit{
+			Owner:    d.Owner,
+			Amount:   sdk.NewUintFromBigInt(d.Amount),
+			BlockNum: sdk.NewUintFromBigInt(d.EthBlockNum),
+		}
+	}
+
+	// save to the db
+	data, err = json.Marshal(deposit)
```

Comment:
Couldn't this be done if the above if statement for when it is not in the cache?

Reply:
that's right